platforms/allwinner-d1/boards/lichee-rv/src/main.rs

```diff
@@ -0,0 +1,263 @@
+};
+use uart::D1Uart;
+mod uart;
+
+const HEAP_SIZE: usize = 384 * 1024 * 1024;
+
+#[link_section = ".aheap.AHEAP"]
+#[used]
+static AHEAP: Ram<HEAP_SIZE> = Ram::new();
+
+#[allow(non_snake_case)]
+#[riscv_rt::entry]
+fn main() -> ! {
+    let mut p = unsafe { d1_pac::Peripherals::steal() };
+    let _uart = unsafe { kernel_uart(&mut p.CCU, &mut p.GPIO, p.UART0) };
+
+    p.GPIO.pc_cfg0.modify(|_r, w| {
+        w.pc1_select().output();
+        w
+    });
+    p.GPIO.pc_dat.modify(|_r, w| {
+        w.pc_dat().variant(0b0000_0010);
+        w
+    });
+
+    // Timer0 is used as a freewheeling rolling timer.
+    // Timer1 is used to generate "sleep until" interrupts
+    //
+    // Both are at a time base of 3M ticks/s.
+    //
+    // In the future, we probably want to rework this to use the RTC timer for
+    // both purposes, as this will likely play better with sleep power usage.
+    let Timers {
+        mut timer0,
+        mut timer1,
+    } = Timers::new(p.TIMER);
+
+    let k = initialize_kernel().unwrap();
+    let dmac = Dmac::new(p.DMAC, &mut p.CCU);
+    let [ch0, ..] = dmac.channels;
+    dmac.dmac.dmac_irq_en0.modify(|_r, w| {
+        w.dma0_queue_irq_en().enabled();
+        w
+    });
+
+    // Initialize LED loop
+    k.initialize(async move {
+        loop {
+            p.GPIO.pc_dat.modify(|_r, w| {
+                w.pc_dat().variant(0b0000_0010);
+                w
+            });
+            k.sleep(Duration::from_millis(250)).await;
+            p.GPIO.pc_dat.modify(|_r, w| {
+                w.pc_dat().variant(0b0000_0000);
+                w
+            });
+            k.sleep(Duration::from_millis(250)).await;
+        }
+    })
+    .unwrap();
+
+    // Initialize SimpleSerial driver
+    k.initialize(async move {
+        D1Uart::register(k, 4096, 4096, ch0).await.unwrap();
+    })
+    .unwrap();
+
+    // Initialize SerialMux
+    k.initialize(async move {
+        loop {
+            // Now, right now this is a little awkward, but what I'm doing here is spawning
+            // a new virtual mux, and configuring it with:
+            // * Up to 16 virtual ports max
+            // * Framed messages up to 512 bytes max each
+            match SerialMux::register(k, 16, 512).await {
+                Ok(_) => break,
+                Err(RegistrationError::SerialPortNotFound) => {
+                    // Uart probably isn't registered yet. Try again in a bit
+                    k.sleep(Duration::from_millis(10)).await;
+                }
+                Err(e) => {
+                    panic!("uhhhh {e:?}");
+                }
+            }
+        }
+    })
+    .unwrap();
+
+    // Loopback on virtual port zero
+    k.initialize(async move {
+        let mut hdl = loop {
+            match SerialMuxHandle::from_registry(k).await {
+                Some(c) => break c,
+                None => {
+                    k.sleep(Duration::from_millis(100)).await;
+                }
+            }
+        };
```

Comment:
nit/TIOLI: this pattern occurs a couple times, we could maybe make a helper for it?

alternatively, we could have the `k.initialize()` call that initializes the serial mux return a joinhandle or something that these init tasks can await, so that we don't need to sleep in a loop? not a blocker in either case, we can make this nicer later.